binderhub/build_local.py

```diff
@@ -0,0 +1,227 @@
+"""
+Contains build of a docker image from a git repository.
+"""
+
+import asyncio
+from collections import defaultdict
+import datetime
+from functools import partial
+import json
+import os
+import threading
+from urllib.parse import urlparse
+import subprocess
+import warnings
+from tornado import ioloop
+
+from tornado.ioloop import IOLoop
+from tornado.log import app_log
+
+from .utils import rendezvous_rank, KUBE_REQUEST_TIMEOUT
+from .build import ProgressEvent, Build
+
+
+# https://github.com/jupyterhub/repo2docker/blob/2021.08.0/repo2docker/utils.py#L13-L58
+def _execute_cmd(cmd, capture=False, **kwargs):
+    """
+    Call given command, yielding output line by line if capture=True.
+
+    Must be yielded from.
+    """
+    if capture:
+        kwargs["stdout"] = subprocess.PIPE
+        kwargs["stderr"] = subprocess.STDOUT
+
+    proc = subprocess.Popen(cmd, **kwargs)
+
+    if not capture:
+        # not capturing output, let subprocesses talk directly to terminal
+        ret = proc.wait()
+        if ret != 0:
+            raise subprocess.CalledProcessError(ret, cmd)
+        return
+
+    # Capture output for logging.
+    # Each line will be yielded as text.
+    # This should behave the same as .readline(), but splits on `\r` OR `\n`,
+    # not just `\n`.
+    buf = []
+
+    def flush():
+        """Flush next line of the buffer"""
+        line = b"".join(buf).decode("utf8", "replace")
+        buf[:] = []
+        return line
+
+    c_last = ""
+    try:
+        for c in iter(partial(proc.stdout.read, 1), b""):
+            if c_last == b"\r" and buf and c != b"\n":
+                yield flush()
+            buf.append(c)
+            if c == b"\n":
+                yield flush()
+            c_last = c
+        if buf:
+            yield flush()
+    finally:
+        ret = proc.wait()
+        if ret != 0:
+            raise subprocess.CalledProcessError(ret, cmd)
+
+
+class LocalRepo2dockerBuild(Build):
+    """Represents a build of a git repository into a docker image.
+
+    This runs a build using the repo2docker command line tool.
+    """
+
+    def __init__(
+        self,
+        q,
+        api,
```

Comment:
Each type of `Build` will probably have its own set of config params - `api` is only for k8s, for example. I'd say that we should probably manage k8s API object as a singleton on the class or something, and so we can stop having to pass that through. Traitlet config specific to k8s can hence also be moved to the k8s builder?

Reply:
I don't think this needs to block this PR though, we can do this in another PR